nntrainer/layers/multi_head_attention_layer.cpp

```diff
@@ -73,10 +73,317 @@ enum AttentionParams {
+    std::get<props::ProjectedKeyDim>(multi_head_attention_props).get();
+
+  if (std::get<props::ProjectedValueDim>(multi_head_attention_props).empty()) {
+    std::get<props::ProjectedValueDim>(multi_head_attention_props)
+      .set(projected_key_dim_prop);
+  }
+
+  if (std::get<props::OutputShape>(multi_head_attention_props).empty()) {
+    std::get<props::OutputShape>(multi_head_attention_props).set(query_width);
+  }
+
+  sm.setActiFunc(ActivationType::ACT_SOFTMAX);
+}
 
 void MultiHeadAttentionLayer::finalize(InitLayerContext &context) {
   finalizeCommon(context);
+
+  const bool disable_bias =
+    std::get<props::DisableBias>(*layer_impl_props).get();
+  auto &weight_initializer =
+    std::get<props::WeightInitializer>(*layer_impl_props).get();
+  auto &bias_initializer = std::get<props::BiasInitializer>(*layer_impl_props);
+  auto &weight_regularizer =
+    std::get<props::WeightRegularizer>(*layer_impl_props);
+  auto &weight_regularizer_constant =
+    std::get<props::WeightRegularizerConstant>(*layer_impl_props);
+  const float &weight_decay =
+    std::get<props::WeightDecay>(*layer_impl_props).get();
+  const float &bias_decay = std::get<props::BiasDecay>(*layer_impl_props).get();
+
+  const unsigned int num_heads =
+    std::get<props::NumHeads>(multi_head_attention_props).get();
+  const unsigned int projected_key_dim_prop =
+    std::get<props::ProjectedKeyDim>(multi_head_attention_props).get();
+  const unsigned int projected_query_dim_prop = projected_key_dim_prop;
+  const unsigned int projected_value_dim_prop =
+    std::get<props::ProjectedValueDim>(multi_head_attention_props).get();
+  const float dropout_rate =
+    std::get<props::DropOutRate>(multi_head_attention_props).get();
+  const bool provide_attention_mask =
+    std::get<props::ProvideAttentionMask>(multi_head_attention_props).get();
+  const std::string return_attention_weight =
+    std::get<props::ReturnAttentionWeight>(multi_head_attention_props).get();
+  const unsigned int output_shape =
+    std::get<props::OutputShape>(multi_head_attention_props).get();
+
+  TensorDim empty_dim;
+
+  const std::vector<TensorDim> &input_dims = context.getInputDimensions();
+  const TensorDim &query_dim = input_dims[inout_idx[INOUT_INDEX::QUERY]];
+  const TensorDim &key_dim = input_dims[inout_idx[INOUT_INDEX::KEY]];
+  const TensorDim &value_dim = input_dims[inout_idx[INOUT_INDEX::VALUE]];
+  const TensorDim &mask_dim = provide_attention_mask
+                                ? input_dims[inout_idx[INOUT_INDEX::MASK]]
+                                : empty_dim;
+
+  const unsigned int batch_size = query_dim.batch();
+  const unsigned int query_height = query_dim.height();
+  const unsigned int query_width = query_dim.width();
+  const unsigned int key_height = key_dim.height();
+  const unsigned int key_width = key_dim.width();
+  const unsigned int value_height = value_dim.height();
+  const unsigned int value_width = value_dim.width();
+
```

Comment:
I think these are duplicated code with in `FinalizeCommon`

Reply:
Done